doc/neps/ufunc-overrides.rst

```diff
@@ -213,83 +370,128 @@ They have indirect interactions, however, because NumPy's
 most numerical classes, the easiest way to override binary operations is
 thus to define ``__array_ufunc__`` and override the corresponding
 Ufunc. The class can then, like :class:`ndarray` itself, define the
-binary operators in terms of Ufuncs. Here, one has to take some care.
-E.g., the simplest implementation would be::
+binary operators in terms of Ufuncs. Here, one has to take some care to
+ensure that one allows for other classes to indicate they are not
+compatible, i.e., implementations should be something like::
 
     class ArrayLike(object):
+        ...
         def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
             ...
             return result
-        ...
-        def __mul__(self, other):
-            return self.__array_ufunc__(np.multiply, '__call__', self, other)
-
-Suppose now, however, that ``other`` is class that does not know how to
-deal with arrays and ufuncs, but does know how to do multiplication::
 
-    class MyObject(object):
-        __array_ufunc__ = None
+        # Option 1: call ufunc directly
         def __mul__(self, other):
-            return 1234
+            if getattr(other, '__array_ufunc__', False) is None:
+                return NotImplemented
+            return np.multiply(self, other)
+
         def __rmul__(self, other):
-            return 4321
+            return np.multiply(other, self)
 
-In this case, standard Python override rules combined with the above
-discussion would imply::
+        def __imul__(self, other):
+            return np.multiply(self, other, out=self)
 
-    mine = MyObject()
-    arr = ArrayLike([0])
+        # Option 2: call into one's own __array_ufunc__
+        def __mul__(self, other):
+            return self.__array_ufunc__(np.multiply, '__call__', self, other)
 
-    mine * arr    # == 1234       OK
-    arr * mine    # TypeError     surprising
+        def __rmul__(self, other):
+            return self.__array_ufunc__(np.multiply, '__call__', other, self)
 
-The reason why this would occur is: because ``MyObject`` is not an
-``ArrayLike`` subclass, Python resolves the expression ``arr * mine`` by
-calling first ``arr.__mul__``. In the above implementation, this would
-just call the Ufunc, which would see that ``mine.__array_ufunc__`` is
-:obj:`None` and raise a :exc:`TypeError`. (Note that if ``MyObject``
-is a subclass of :class:`ndarray`, Python calls ``mine.__rmul__`` first.)
+        def __imul__(self, other):
+            result = self.__array_ufunc__(np.multiply, '__call__', self, other,
+                                          out=self)
```

Comment:
Nice catches; here it is even wrong, since for `__array_ufunc__` it really should be a tuple.